### PR TITLE
Sectioned selector view controllers

### LINKED ISSUE #871: Divide PushRow with sections
Is it possible to divide the given options into multiple sections?

### PR DESCRIPTION
**Pull request by ilyapuchka**
Added `SectionedPushRow` and `SectionedMultipleSelectorRow` to be able to display selection options in sections grouped by some generic key. 
Resolves #871

### CODE REVIEW

**Comment by ilyapuchka** [comment]
Any feedback on that?

**Comment by mtnbarreto** [comment]
Hi @ilyapuchka Thanks for implementing such a great feature!
Overall I liked the code style and the abstraction used.
I'm gonna add my feedback as a pull request review shortly. Sorry for the delay on that.

File: Source/Rows/Controllers/MultipleSelectorViewController.swift [hunk 1]
```diff
@@ -67,25 +77,54 @@ open class _MultipleSelectorViewController<T:Hashable, Row: SelectableRowType> :
...
+            }
+        }
+        return section
+    }
+}
+
+open class _SectionedMultipleSelectorViewController<K: Hashable & Comparable, T: Hashable, Row: SelectableRowType>: _MultipleSelectorViewController<T, Row> where Row: BaseRow, Row: TypedRowType, Row.Cell.Value == T {
+    
+    public var sectionKeyFor: ((Row.Cell.Value) -> (K))?
+    public var sectionHeaderTitleFor: ((K) -> String?)? = { String(describing: $0) }
```
Comment (mtnbarreto) [inline Source/Rows/Controllers/MultipleSelectorViewController.swift]:
```swift
public var sectionHeaderTitleFor: ((String) -> HeaderFooterViewRepresentable?)? = { HeaderFooterView<UITableViewHeaderFooterView>(title: String(describing: $0))  }
```

File: Source/Rows/Controllers/MultipleSelectorViewController.swift [hunk 2]
```diff
@@ -67,25 +77,54 @@ open class _MultipleSelectorViewController<T:Hashable, Row: SelectableRowType> :
...
+    public var sectionFooterTitleFor: ((K) -> String?)?
```
Comment (mtnbarreto) [inline Source/Rows/Controllers/MultipleSelectorViewController.swift]: same change should be applied here

File: Example/Example/ViewController.swift [hunk 3]
```diff
@@ -219,7 +219,23 @@ class RowsExampleViewController: FormViewController {
                         $0.value = 👦🏼
                         $0.selectorTitle = "Choose an Emoji!"
                     }
-            
+
+                <<< SectionedPushRow<String, Emoji>() {
```
Comment (mtnbarreto) [inline Example/Example/ViewController.swift]: I would prefer to use Strings always as key type (K generic type) so we don't need to specify it.

File: Source/Rows/Controllers/SelectorViewController.swift [hunk 4]
```diff
@@ -56,32 +71,56 @@ open class _SelectorViewController<Row: SelectableRowType>: FormViewController,
...
-/// Selector Controller (used to select one option among a list)
-open class SelectorViewController<T:Equatable> : _SelectorViewController<ListCheckRow<T>>  {
+open class _SectionedSelectorViewController<K: Hashable & Comparable, Row: SelectableRowType>: _SelectorViewController<Row> where Row: BaseRow, Row: TypedRowType {
     
-    override public init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
-        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
-    }
+    public var sectionKeyFor: ((Row.Cell.Value) -> (K))?
+    public var sectionHeaderTitleFor: ((K) -> String?)? = { String(describing: $0) }
+    public var sectionFooterTitleFor: ((K) -> String?)?
```
Comment (mtnbarreto) [inline Source/Rows/Controllers/SelectorViewController.swift]: Please move these closure to _SelectorViewController type, replace K by String.

File: Source/Rows/Controllers/SelectorViewController.swift [hunk 5]
```diff
@@ -56,32 +71,56 @@ open class _SelectorViewController<Row: SelectableRowType>: FormViewController,
...
     }
     
+}
 
+/// Selector Controller (used to select one option among a list)
+open class SelectorViewController<T:Equatable> : _SelectorViewController<ListCheckRow<T>>  {
+}
+
+/// Selector Controller (used to select one option among a list)
+open class SectionedSelectorViewController<K: Hashable & Comparable, T:Equatable> : _SectionedSelectorViewController<K, ListCheckRow<T>>  {
```
Comment (mtnbarreto) [inline Source/Rows/Controllers/SelectorViewController.swift]: Delete `Sectioned....` types..

**Comment by mtnbarreto** [comment]
If you are ok with the changes please rebase your code with master branch before updating the code.

File: Source/Core/SelectableSection.swift [hunk 6]
```diff
@@ -132,7 +132,12 @@ open class SelectableSection<Row: SelectableRowType> : Section, SelectableSectio
         self.selectionType = selectionType
         super.init(header, initializer)
     }
-    
+
+    public init(header: String, footer: String, selectionType: SelectionType, _ initializer: (Section) -> () = { _ in }) {
```
Comment (ilyapuchka) [inline Source/Core/SelectableSection.swift]: It will be still possible to use custom headers by setting them manually as section property, like described in the link that you shared. We are doing exactly that in our app. 
Currently `Section` initialisers are not providing `HeaderFooterViewRepresentable` as parameter, only `String`. It will make sense to improve that but I believe it should be done in a separate PR.

**Comment by ilyapuchka** [comment]
@mtnbarreto thanks for review, I addressed your comments and simplified code by removing subclassing.

File: Source/Core/SelectableSection.swift (hunk 6 above)
Comment (mtnbarreto) [inline Source/Core/SelectableSection.swift]: Agreed, let's do it in another pull request..

File: Example/Example/ViewController.swift (hunk 3 above)
Comment (mtnbarreto) [inline Example/Example/ViewController.swift]: What about this?

Comment (ilyapuchka) [inline Example/Example/ViewController.swift]: This is now removed, now it's just a normal PushRow and you can set closures to break options by sections in `onPresent` block

Comment (mtnbarreto) [inline Example/Example/ViewController.swift]: sorry i haven't seen the new commit.

**Comment by mtnbarreto** [comment]
Thanks @ilyapuchka , now it looks great! Let's merge it.

**Comment by OxyFlax** [comment]
Could you please explain what is possible thanks to this PR? I haven't seen SectionedPushRow or any other option to do it in the documentation and couldn't find it in PushRow class.

**Comment by mats-claassen** [comment]
You can show the `PushRow` options splitted in sections using the `sectionKeyForValue` block.

You can see an example of this in the Example project. Have a look at the `RowsExample` view controller.
